File: csharp/Creational/03-Builder/Concrete Builder/JobCardBuilder.cs
public class JobCardBuilder : IJobCardBuilder
{
    private readonly string _title;
    private readonly string _category;
    private string          _priority           = "NORMAL";
    private DateTime?       _scheduledAt;
    private List<string>    _parts              = new();
    private List<string>    _photoUrls          = new();
    private string?         _instructions;
    private bool            _requiresInspection;
    private bool            _hasWarranty;

    public JobCardBuilder(string title, string category)
    {
        _title    = title;
        _category = category;
    }

    public IJobCardBuilder WithPriority(string priority)
    {
        _priority = priority;
        return this;
    }

    public IJobCardBuilder ScheduledFor(DateTime dateTime)
    {
        _scheduledAt = dateTime;
        return this;
    }

    public IJobCardBuilder WithParts(params string[] parts)
    {
        _parts.AddRange(parts);
        return this;
    }

    public IJobCardBuilder WithPhotos(params string[] urls)
    {
        _photoUrls.AddRange(urls);
        return this;
    }

    public IJobCardBuilder WithInstructions(string instructions)
    {
        _instructions = instructions;
        return this;
    }

    public IJobCardBuilder RequiresSafetyInspection()
    {
        _requiresInspection = true;
        return this;
    }

    public IJobCardBuilder WithWarranty()
    {
        _hasWarranty = true;
        return this;
    }

    public JobCard Build()
    {
        if (string.IsNullOrWhiteSpace(_title))
            throw new InvalidOperationException("Job card must have a title.");

        return new JobCard
        {
            Title               = _title,
            Category            = _category,
            Priority            = _priority,
            ScheduledAt         = _scheduledAt,
            Parts               = _parts,
            PhotoUrls           = _photoUrls,
            SpecialInstructions = _instructions,
            RequiresInspection  = _requiresInspection,
            HasWarranty         = _hasWarranty,
        };
    }
}

File: typescript/Creational/03-Builder/Concrete Builder/JobCardBuilder.ts
import { JobCard } from "../Product/JobCard";
import type { IJobCardBuilder } from "../Builder Interface/IJobCardBuilder";

export class JobCardBuilder implements IJobCardBuilder {
  private _priority            = "NORMAL";
  private _scheduledAt?:       Date;
  private _parts:              string[] = [];
  private _photoUrls:          string[] = [];
  private _instructions?:      string;
  private _requiresInspection  = false;
  private _hasWarranty         = false;

  constructor(
    private readonly _title:    string,
    private readonly _category: string
  ) {}

  withPriority(priority: string):  IJobCardBuilder { this._priority = priority;           return this; }
  scheduledFor(date: Date):        IJobCardBuilder { this._scheduledAt = date;             return this; }
  withParts(...parts: string[]):   IJobCardBuilder { this._parts.push(...parts);           return this; }
  withPhotos(...urls: string[]):   IJobCardBuilder { this._photoUrls.push(...urls);        return this; }
  withInstructions(text: string):  IJobCardBuilder { this._instructions = text;            return this; }
  requiresSafetyInspection():      IJobCardBuilder { this._requiresInspection = true;      return this; }
  withWarranty():                  IJobCardBuilder { this._hasWarranty = true;             return this; }

  build(): JobCard {
    if (!this._title.trim())
      throw new Error("Job card must have a title.");

    return new JobCard({
      title:               this._title,
      category:            this._category,
      priority:            this._priority,
      scheduledAt:         this._scheduledAt,
      parts:               this._parts,
      photoUrls:           this._photoUrls,
      specialInstructions: this._instructions,
      requiresInspection:  this._requiresInspection,
      hasWarranty:         this._hasWarranty,
    });
  }
}